### PR TITLE
Bug fix and testcases for edit command

### PR DESCRIPTION
**Pull request by CharlesLee01**
Same as PR #281 just with some updates so that it fits the recent changes in typicalId and typicalEntities. Some bug fixes due to overwritten function not remove despite inherited from parent.

### CODE REVIEW

File: src/main/java/dog/pawbook/logic/commands/EditDogCommand.java [hunk 1]
```diff
@@ -52,36 +52,6 @@ public EditDogCommand(Integer id, EditDogDescriptor editDogDescriptor) {
...
-        // special handling of Owner ID change
-        int originalOwnerId = targetDog.getOwnerId();
-        int editedOwnerId = editedDog.getOwnerId();
-        if (originalOwnerId != editedOwnerId) {
-            changeOwner(model, originalOwnerId, editedOwnerId);
-        }
-
-        model.setEntity(id, editedDog);
-        return new CommandResult(getSuccessMessage(editedDog));
-    }
```
Comment (kouyk) [inline src/main/java/dog/pawbook/logic/commands/EditDogCommand.java:83]: Do you mind explaining why this is taken away? I cannot seem to find a reason that the override should not happen.

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        String expectedMessage = String.format(EditDogCommand.MESSAGE_EDIT_DOG_SUCCESS, editedDog);
+
+        Model expectedModel = new ModelManager(new Database(model.getDatabase()), new UserPrefs());
+        expectedModel.setEntity(ID_TWO, editedDog);
+        expectedModel.updateFilteredEntityList(new IdMatchPredicate(ID_TWO));
+
+        assertCommandSuccess(editEntityCommand, model, expectedMessage, expectedModel);
+    }
+
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java:74]: add a test that changes the owner ID, and also one that causes a duplicate, change the other 2 accordingly as well as you deem fit.

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java [hunk 3]
```diff
@@ -36,7 +36,8 @@ public void equals() {
         assertFalse(DESC_AMY.equals(DESC_BOB));
 
         // different name -> returns false
-        EditOwnerDescriptor editedAmy = new EditOwnerDescriptorBuilder(DESC_AMY).withName(VALID_NAME_BOB).build();
+        EditOwnerDescriptor editedAmy = (EditOwnerDescriptor) new EditOwnerDescriptorBuilder(DESC_AMY)
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:39]: do you mind explaining why casting is necessary?

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java [hunk 4]
```diff
@@ -52,7 +53,7 @@ public void equals() {
         assertFalse(DESC_AMY.equals(editedAmy));
 
         // different tags -> returns false
-        editedAmy = new EditOwnerDescriptorBuilder(DESC_AMY).withTags(VALID_TAG_HUSBAND).build();
+        editedAmy = (EditOwnerDescriptor) new EditOwnerDescriptorBuilder(DESC_AMY).withTags(VALID_TAG_HUSBAND).build();
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:56]: and this too

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        // no field specified
+        assertParseFailure(parser, "1", EditDogCommand.MESSAGE_NOT_EDITED);
+
+        // no index and no field specified
+        assertParseFailure(parser, "", MESSAGE_INVALID_FORMAT);
+    }
+
+    @Test
+    public void parse_invalidPreamble_failure() {
+        // negative index
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:68]: it is ID now

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        assertParseFailure(parser, "-5" + NAME_DESC_ASHER, MESSAGE_INVALID_FORMAT);
+
+        // zero index
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:71]: here also

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        assertParseFailure(parser, "1" + INVALID_DATEOFBIRTH_DESC, DateOfBirth.MESSAGE_CONSTRAINTS);
+
+        // invalid breed followed by valid dob
+        assertParseFailure(parser, "1" + INVALID_BREED_DESC + DATEOFBIRTH_DESC_ASHER, Breed.MESSAGE_CONSTRAINTS);
+
+        // valid breed followed by invalid breed. The test case for invalid breed followed by valid breed
+        // is tested at {@code parse_invalidValueFollowedByValidValue_success()}
+        assertParseFailure(parser, "1" + BREED_DESC_BELL + INVALID_BREED_DESC, Breed.MESSAGE_CONSTRAINTS);
+
+        // while parsing {@code PREFIX_TAG} alone will reset the tags of the {@code Owner} being edited,
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:97]: check the javadoc!

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        // parsing it together with a valid tag results in error
+        assertParseFailure(parser, "1" + TAG_DESC_FRIENDLY + TAG_DESC_QUIET + TAG_EMPTY, Tag.MESSAGE_CONSTRAINTS);
+        assertParseFailure(parser, "1" + TAG_DESC_FRIENDLY + TAG_EMPTY + TAG_DESC_QUIET, Tag.MESSAGE_CONSTRAINTS);
+        assertParseFailure(parser, "1" + TAG_EMPTY + TAG_DESC_FRIENDLY + TAG_DESC_QUIET, Tag.MESSAGE_CONSTRAINTS);
+
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:102]: how about invalid values for the other attribute fields like sex?

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        // multiple invalid values, but only the first invalid value is captured
+        assertParseFailure(parser, "1" + INVALID_NAME_DESC + INVALID_BREED_DESC + VALID_DATEOFBIRTH_ASHER
+                + VALID_SEX_BELL, Name.MESSAGE_CONSTRAINTS);
+    }
+
+    @Test
+    public void parse_allFieldsSpecified_success() {
+        Integer targetId = ID_TWO;
+        String userInput = targetId + BREED_DESC_ASHER + TAG_DESC_FRIENDLY
+                + DATEOFBIRTH_DESC_ASHER + SEX_DESC_ASHER + NAME_DESC_ASHER + TAG_DESC_QUIET;
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:112]: are you sure this is all that's editable?

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        userInput = targetId + SEX_DESC_ASHER;
+        descriptor = new EditDogDescriptorBuilder().withSex(VALID_SEX_ASHER).build();
+        expectedCommand = new EditDogCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+
+        // tags
+        userInput = targetId + TAG_DESC_FRIENDLY;
+        descriptor = new EditDogDescriptorBuilder().withTags(VALID_TAG_FRIENDLY).build();
+        expectedCommand = new EditDogCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:165]: owner?

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        // no index and no field specified
+        assertParseFailure(parser, "", MESSAGE_INVALID_FORMAT);
+    }
+
+    @Test
+    public void parse_invalidPreamble_failure() {
+        // negative index
+        assertParseFailure(parser, "-5" + NAME_DESC_OBEDIENCE_TRAINING, MESSAGE_INVALID_FORMAT);
+
+        // zero index
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:59]: see my comments for the dog version

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        EditProgramDescriptor descriptor = new EditProgramDescriptorBuilder()
+                .withSessions(VALID_SESSION_POTTY_TRAINING)
+                .withTags(VALID_TAG_ALL, VALID_TAG_PUPPIES).build();
+        EditProgramCommand expectedCommand = new EditProgramCommand(targetId, descriptor);
+
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_resetTags_success() {
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:157]: you need a resetSessions equivalent

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        descriptor = new EditProgramDescriptorBuilder().withTags(VALID_TAG_ALL).build();
+        expectedCommand = new EditProgramCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_multipleRepeatedFields_acceptsLast() {
+        Integer targetId = ID_ONE;
+        String userInput = targetId + TAG_DESC_PUPPIES + TAG_DESC_ALL + TAG_DESC_ALL + TAG_DESC_PUPPIES
+                + SESSION_DESC_POTTY_TRAINING + SESSION_DESC_POTTY_TRAINING + TAG_DESC_PUPPIES;
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:146]: I think we need another test that ensures that those repeatable fields like tags etc, if different values are supplied, all will be captured, and if repeated, it will be deduplicated correctly.

**Review by kouyk — CHANGES_REQUESTED** [review]
There are several areas for improvement to make the test case more comprehensive, my main gripe is that you closed the previous PR and made a new one, something that you have done in the past which was still acceptable when the changes are small, but with changes this huge, it is difficult to keep track of where we last left off especially when you had some delays in updating the PR. Another thing is that some of the authorship is lost in the process, like the parts that I fixed your builder patterns such that there is no longer unsafe type casts.

**Comment by kouyk** [comment]
I forgot another issue I want to bring up, while writing test case will let us discover bugs, it doesn't always mean that something that doesn't work with your test case is wrong, it might be that your test case is written wrongly in the first place, before modifying the functional code, please double check or even do some manual testing to see if the bug can be replicated to confirm that the behaviour is wrong...

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java (hunk 3 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:39]: I just copy n paste from the previous PR. This part was accepted last time so I assume it works for now?

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java (hunk 4 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:56]: Same reason as above

**Comment by CharlesLee01** [comment]
Can I know how to pull this PR to my local repo? Coz I made some changes here directly.

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java (hunk 3 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:39]: I don't want to do any finger-pointing here, while this is my oversight the previous time, it would be helpful if you broke down your commits such that every commit has a purposeful message along with the stated changes, that way it is easier for me look through at each step and figure out if anything is wrong. However, all that I get here is around 1k lines of code changed, barring the fact that the previous commit history didn't carry over, it is really difficult to be that careful. I won't request that you review what I changed for you in the previous PR, in short your builder pattern was less than ideal, thus sub-type information was lost, hence requiring the casting. I fixed it to avoid those issues, but since extra casting is not technically wrong but just poor coding practice, there wasn't anything that alerted or reminded me that these portions needed to be updated to remove the casting. The fix here is simple, just remove the cast. Do let me know if you don't understand any part.

File: src/main/java/dog/pawbook/logic/commands/EditDogCommand.java (hunk 1 above)
Comment (CharlesLee01) [reply src/main/java/dog/pawbook/logic/commands/EditDogCommand.java:83]: I think this part is causing some bugs. Coz from there's no execute() in editOwnerCommand and editProgramCommand. So this execute() should inherit from editDogCommand as well since it already inherited the editEntityCommand.

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java (hunk 3 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:39]: Remove!

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java (hunk 4 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:56]: Remove!

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java (hunk 8 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:102]: Other fields are at the top right?

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java (hunk 13 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:146]: Wait, I thought this clearly covers the duplicated tags and different tags? I can add different sessions if that will help.

File: src/main/java/dog/pawbook/logic/commands/EditDogCommand.java (hunk 1 above)
Comment (kouyk) [reply src/main/java/dog/pawbook/logic/commands/EditDogCommand.java:83]: Restore it back, it is essential, or else when the owner ID of a dog changes, the old owner will still have the dog

File: src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java (hunk 3 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/commands/EditOwnerDescriptorTest.java:39]: It is not just whether or not someone is going to review it, git commits are supposed to be incremental regardless, the previous PR could have been kept by merging in the latest master and resolving the conflicts. By creating another PR the commit history and authorship are all gone, that's the same reason I previously tried to restrain myself from directly pushing to existing PRs and make everyone perform the minor changes themselves, it actually takes more effort to type all the comments I have given than to directly modify them myself.

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java (hunk 13 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:146]: I'm referring to more test cases that can be written, this line is the most logical thing I can point to. For tags and sessions, repeated values are all captured, but the tests clearly don't look out for this, in this current test case there is only repeated tags, which will be parsed as one.

**Comment by CharlesLee01** [comment]
Made those changes! Just that after restoring the execute() there's one test case failing in editDogCommandTest which I have commented out. I need your help for that, I have tested the database.equals(other.database) in equals() of ModelManager is failing. I tried printing it, the database is exactly the same object but it fails...

**Comment by CharlesLee01** [comment]
Also, is possible to merge this PR coz it's getting bigger

**Comment by kouyk** [comment]
To save you the hassle, the mistake you've made is forgetting to delete the dog ID its owner.  ---- On Fri, 09 Apr 2021 18:52:41 +0800  ***@***.***  wrote ----
Made those changes! Just that after restoring the execute() there's one test case failing in editDogCommandTest which I have commented out. I need your help for that, I have tested the database.equals(other.database) in equals() of ModelManager is failing. I tried printing it, the database is exactly the same object but it fails...

**Comment by kouyk** [comment]
It's fine, breaking down PR is for every PR to serve a single purpose, I rather not have a single PR for every "minor change" or "some change" that you have made. This PR dragged long enough that a little while longer won't hurt ---- On Fri, 09 Apr 2021 18:54:45 +0800  ***@***.***  wrote ----
Also, is possible to merge this PR? coz it's getting bigger

**Comment by CharlesLee01** [comment]
This is what I have done so far. I added the part for delete dogid from old owner but it still fails. Do you mind helping me with this part, I'm really stucked. Perhaps, I can start working on developer guide.

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java (hunk 2 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java:79]: To avoid any confusion, I meant to **add** an additional test for this, not to rename this test method.

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java (hunk 8 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:102]: my bad, do standardise how the comments are placed for the block of code, if i'm making this mistake it means the assessors might also make similar mistakes, where you won't be able to explain yourself anymore

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        expectedCommand = new EditProgramCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_multipleRepeatedFields_acceptsLast() {
+        Integer targetId = ID_ONE;
+        String userInput = targetId + TAG_DESC_PUPPIES + SESSION_DESC_OBEDIENCE_TRAINING + TAG_DESC_ALL
+                + TAG_DESC_ALL + TAG_DESC_PUPPIES + SESSION_DESC_POTTY_TRAINING
+                + SESSION_DESC_POTTY_TRAINING + TAG_DESC_PUPPIES + SESSION_DESC_OBEDIENCE_TRAINING;
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:148]: the method name implies that multiple fields that should only be specified once are now specified multiple times, in which only the last repetition is to be accepted as the final value, programs clearly don't have such a field when being edited, both tags and sessions accepts all of the fields and will only ignore repeated values, what I need from you is to write test methods that instead test if multiple tags & sessions are all accepted.

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        Model expectedModel = new ModelManager(new Database(model.getDatabase()), new UserPrefs());
+        expectedModel.setEntity(ID_TWO, editedDog);
+        expectedModel.updateFilteredEntityList(new IdMatchPredicate(ID_TWO));
+
+        Owner toEditOwner = (Owner) model.getEntity(ID_ONE);
+        Set<Integer> newDogIdSet = new HashSet<>(toEditOwner.getDogIdSet());
+        newDogIdSet.remove(ID_TWO);
+        model.setEntity(ID_ONE, new Owner(toEditOwner.getName(), toEditOwner.getPhone(), toEditOwner.getEmail(),
+                toEditOwner.getAddress(), toEditOwner.getTags(), newDogIdSet));
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java:101]: Also, in the copy-paste process, you might have not spent the time to understand what each of the test are doing, there is an `expectedModel` which we modify by hand in the most crude way to keep things simple and avoid errors, while the `model` is the target that we let the `EditDogCommand` execute on and therefore modify.

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+        Model expectedModel = new ModelManager(new Database(model.getDatabase()), new UserPrefs());
+        expectedModel.setEntity(ID_TWO, editedDog);
+        expectedModel.updateFilteredEntityList(new IdMatchPredicate(ID_TWO));
+
+        assertCommandSuccess(editEntityCommand, model, expectedMessage, expectedModel);
+    }
+
+    /*@Test
+    public void execute_someFieldsSpecified_success() {
```
Comment (kouyk) [inline src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java:81]: Just in case you don't understand how Pawbook works, I will explain it here in the interest of everyone's time. Every dog has an owner, and so long as the dog exists, the owner must also exist, but not the other way around, so owners can be dogless and still exist. While we don't allow for dogs to be created without a valid owner, it is possible to change the ownership afterwards. If the edit requests for dog with ID 2 to be owned by owner with ID 3, obviously the original owner with ID 1 now needs to disown the dog, followed by owner with ID 3 taking over the dog instead. In case you're still confused, 3 entities have been edited at this point, 2 owners 1 dog.

**Review by kouyk — COMMENTED** [review]
I hope my comments explains it well enough for you to proceed and fix these issues asap.

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java (hunk 14 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:148]: > what I need from you is to write test methods that instead test if multiple tags & sessions are all accepted.
You mean to test multiple tags and sessions?

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java (hunk 16 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java:81]: Um, I know how Pawbook works lol. Instead of explaining how the function work, can you specify which part the error is caused? I tried adding dogId to new owner and remove dogId from the old owner before I pushed the latest commit here but it doesn't work as well. I'll push the one I did which follows the approach u mentioned here, but it has some assertion error which highlights that the owner to remove dog id doesn't have that dog id. You can have a look at it.

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java (hunk 13 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java]: Done!

File: src/main/java/dog/pawbook/logic/commands/EditDogCommand.java (hunk 1 above)
Comment (CharlesLee01) [reply src/main/java/dog/pawbook/logic/commands/EditDogCommand.java]: Done!

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java (hunk 2 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java:85]: Done!

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java (hunk 14 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java:148]: Yes, multiple valid but different tags and sessions, ensure that not only the last is accepted.

Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java]: Done! Tell me if it's still not enough

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java (hunk 16 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java]: First mistake: you copied code from the Class you're testing against mixed in with code of your own, which also contains helper methods that triggered the assertion errors
Second mistake: you have been performing the changes on both the expectedModel and the original model, not sure what you're expecting out of this

Comment (kouyk) [reply src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java]: > can you specify which part the error is caused?

I have already pointed out the second mistake in another comment, not sure what you meant

File: src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java (hunk 14 above)
Comment (kouyk) [reply src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java]: > what I need from you is to write test methods that instead test if multiple tags & sessions are all accepted

This refers to additional test methods, the current test method in which you added additional stuff is pointless, keep it as the original is fine. Write **another** test method that specifically test multiple valid and different tags overriding the existing ones, and another that test multiple valid and different session timings to see if it overrides the existing ones

File: src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java (hunk 16 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java]: Done!

File: src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java (hunk 8 above)
Comment (CharlesLee01) [reply src/test/java/dog/pawbook/logic/parser/EditDogCommandParserTest.java:103]: Done!

**Review by kouyk — APPROVED** [review]
